src/AWS.Deploy.Orchestration/CustomRecipeLocator.cs

```diff
@@ -0,0 +1,177 @@
+        /// <returns>A list of recipe definition paths.</returns>
+        private List<string> GetRecipePathsFromRootDirectory(string? rootDirectoryPath)
+        {
+            var recipePaths = new List<string>();
+            if (!string.IsNullOrEmpty(rootDirectoryPath) && _directoryManager.Exists(rootDirectoryPath))
+            {
+                foreach (var recipeFilePath in _directoryManager.GetFiles(rootDirectoryPath, "*.recipe", SearchOption.AllDirectories))
+                {
+                    if (recipeFilePath.Contains(_ignorePathSubstring))
+                        continue;
+                    recipePaths.Add(_directoryManager.GetDirectoryInfo(recipeFilePath).Parent.FullName);
+                }
+            }
+            return recipePaths;
+        }
+
+        /// <summary>
+        /// This method finds the root directory that is monitored by the same source control as the current directory.
+        /// </summary>
+        /// <param name="currentDirectoryPath">The absolute path of the current directory</param>
+        /// <returns>The source control root directory absolute path.</returns>
+        private async Task<string?> GetSourceControlRootDirectory(string currentDirectoryPath)
+        {
+            var possibleRootDirectoryPath = currentDirectoryPath;
+            while (currentDirectoryPath != null && await IsDirectoryUnderSourceControl(currentDirectoryPath))
+            {
+                possibleRootDirectoryPath = currentDirectoryPath;
+                currentDirectoryPath = _directoryManager.GetDirectoryInfo(currentDirectoryPath).Parent.FullName;
+            }
+            return possibleRootDirectoryPath;
+        }
+
+        /// <summary>
+        /// Helper method to find if the directory is monitored by a source control system.
+        /// </summary>
+        /// <param name="directoryPath">An absolute directory path.</param>
+        /// <returns></returns>
+        private async Task<bool> IsDirectoryUnderSourceControl(string? directoryPath)
+        {
+            if (!string.IsNullOrEmpty(directoryPath))
+            {
+                var gitStatusResult = await _commandLineWrapper.TryRunWithResult(GIT_STATUS_COMMAND, directoryPath);
+                var svnStatusResult = await _commandLineWrapper.TryRunWithResult(SVN_STATUS_COMMAND, directoryPath);
+                return gitStatusResult.Success || svnStatusResult.Success;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// This method determines if the given directory contains any recipe files
+        /// </summary>
+        /// <param name="directoryPath">The path of the directory that needs to be validated</param>
+        /// <returns>A bool indicating the presence of a recipe file inside the directory.</returns>
+        private bool ContainsRecipeFile(string directoryPath)
+        {
+            return _directoryManager.GetFiles(directoryPath, "*.recipe").Length > 0;
```

Comment:
deploy tool will not work with any recipe, it must be either with an explicit name or constant.

we agreed on `app.recipe` convention last time.

Reply:
Added a filter to only include recipe files that have the same name as the parent directory